### PR TITLE
New Adapter: Displayio

### PR DESCRIPTION
**Pull request by xdevel**
Display.io prebid server adapter

Docs PR: https://github.com/prebid/prebid.github.io/pull/5330

### CODE REVIEW

File: static/bidder-info/displayio.yaml
Comment (onkarvhanumante) [inline static/bidder-info/displayio.yaml:1]: consider that usage of partial dynamic urls is being discouraged and prebid team in future will be working to avoid usage of dynamic subdomain in endpoint url.

Major concerns with such usage are,

security concerns
The security aspect is alleviated by using a fixed top level domain. Due to the potential harm to hosts, we are strict in this requirement. We are working towards fixing the few adapters currently in violation.

connection performance
The connection performance advice is for your benefit. Client specific subdomains prevent Prebid Server from reusing connections across your clients which results in more connections needed for your adapter. The issue gets worse the more successful you are in attracting new clients. Hosts may choose to disable your adapter due to this behavior.

**Comment by onkarvhanumante** [comment]
@xdevel 
should update json test
- to have request with multiple imps
- to have multi-format imp request

File: adapters/displayio/displayio.go
Comment (onkarvhanumante) [inline adapters/displayio/displayio.go:38]: PBS core framework has this check. Adapter code doesn't need to check for it

File: adapters/displayio/displayio.go
Comment (onkarvhanumante) [inline adapters/displayio/displayio.go:106]: Based on adapter code, looks like it is expected to have same placementId and inventoryId for all imps. @xdevel could you confirm if this is expected behaviour. If yes then should specify this in bidder docs.

File: static/bidder-info/displayio.yaml
Comment (xdevel) [reply static/bidder-info/displayio.yaml]: fixed

File: adapters/displayio/displayio.go
Comment (xdevel) [reply adapters/displayio/displayio.go]: @onkarvhanumante No it's not expected nor required. This code part is rewritten to have separate placement/inventory ids per imp. Thanks.

File: adapters/displayio/displayio.go
Comment (xdevel) [reply adapters/displayio/displayio.go]: Ok, removed

File: static/bidder-info/displayio.yaml
Comment (onkarvhanumante) [inline static/bidder-info/displayio.yaml:1]: endpoint is reachable

```
% curl -i --location --request POST 'https://prebid.display.io/?publisher=1234'
HTTP/1.1 411 Length Required
content-type: text/html
content-length: 353
date: Tue, 11 Jun 2024 05:30:20 GMT
server: nginx/1.14.0
```

File: static/bidder-info/displayio.yaml
Comment (onkarvhanumante) [inline static/bidder-info/displayio.yaml:6]: Prebid team has sent email to verify above maintainers email. Requesting to responding back on email thread

File: adapters/displayio/displayio_test.go
Comment (onkarvhanumante) [inline adapters/displayio/displayio_test.go:14]: should use endpoint specified in `displayio.yaml` - `https://prebid.display.io/?publisher={{.PublisherID}}`

this ensures adapter code works correctly to substitute macro

**Comment by onkarvhanumante** [comment]
@xdevel should update json test
- to have request with multiple imps
- to have multi-format imp request

File: adapters/displayio/displayio_test.go
Comment (xdevel) [reply adapters/displayio/displayio_test.go]: Endpoint updated

**Comment by xdevel** [comment]
@onkarvhanumante both tests are already added, please take a look on multi-format.json / multi-imp.json

**Comment by rgushel** [comment]
@onkarvhanumante any updates on this one?

File: static/bidder-info/displayio.yaml
Comment (onkarvhanumante) [reply static/bidder-info/displayio.yaml:6]: received response from `contact@display.io`

File: adapters/displayio/displayio.go
Comment (onkarvhanumante) [inline adapters/displayio/displayio.go:94]: any particular reason to unMarshal  `request.Ext` on line 89?  unMarshalled value is always overwritten by `dioExt` assignment on line 94

File: adapters/displayio/displayio.go
Comment (onkarvhanumante) [inline adapters/displayio/displayio.go:79]: no need to add this check. Prebid core code will make sure Bidder value is present

Comment (xdevel) [reply adapters/displayio/displayio.go]: ok, removed

File: adapters/displayio/displayio.go
Comment (xdevel) [reply adapters/displayio/displayio.go:91]: @onkarvhanumante it's not overwritten but modified. We alter request.ext with dio-related data.

**Review by onkarvhanumante — APPROVED** [review]

File: adapters/displayio/displayio_test.go
Comment (xdevel) [reply adapters/displayio/displayio_test.go]: URLs updated

File: adapters/displayio/displayio.go
Comment (xdevel) [reply adapters/displayio/displayio.go]: Imports reorganized

File: adapters/displayio/displayio.go
Comment (xdevel) [reply adapters/displayio/displayio.go:28]: Ok, thanks for suggestion. I changed the code to use a copy of request for each imp

File: adapters/displayio/displayiotest/exemplary/simple-banner.json
Comment (xdevel) [reply adapters/displayio/displayiotest/exemplary/simple-banner.json:1]: Tests simplified. I didn't dive a lot into it but tried to clear some unnecessary fields

File: adapters/displayio/displayiotest/exemplary/multi-imp.json
Comment (xdevel) [reply adapters/displayio/displayiotest/exemplary/multi-imp.json]: Ok, updated

File: adapters/displayio/displayiotest/exemplary/multi-format.json
Comment (xdevel) [reply adapters/displayio/displayiotest/exemplary/multi-format.json]: >Is it typical for this information to exist in the request.ext of the incoming request?
No, no matter what will be send under `request.ext.displayio`, it will be replaced by adapter code.
I dropped it from incoming request tests to keep it straight

File: adapters/displayio/displayiotest/supplemental/missing-bidfloorcur.json
Comment (xdevel) [reply adapters/displayio/displayiotest/supplemental/missing-bidfloorcur.json:1]: You're right. File renamed

**Review by bsardo — APPROVED** [review]

**Comment by rgushel** [comment]
@onkarvhanumante looking forward to your final check. Thanks!

**Comment by rgushel** [comment]
@onkarvhanumante we've fixed all the blockers and are waiting for your approval

**Review by onkarvhanumante — APPROVED** [review]

**Comment by onkarvhanumante** [comment]
> @onkarvhanumante we've fixed all the blockers and are waiting for your approval

Thanks for being patient with PR reviews. PR is merged

**Pull request merged**
